File: Week 29/rust/caching.rs
// WEEK 29 - RUST ADVANCED TOPICS
// Topic: Caching - LRU, LFU and TTL
// File: caching.rs
//
// CONCEPT:
//   Caches accelerate access to recurring queries by storing values in
//   limited fast storage. Eviction policy decides what to drop:
//     - LRU: drop least recently used.
//     - LFU: drop least frequently used.
//     - TTL: drop after a deadline.
//
// KEY POINTS:
//   - LRU: ordered map of (key, val); on touch move to back.
//   - LFU: HashMap of (val, freq, list iter) + freq -> VecDeque<key>.
//   - TTL: HashMap<K, (V, Instant)>; lazy eviction on access.
//
// ALGORITHM / APPROACH:
//   LRU touch: remove(key); insert at back. On overflow drop front.
//   LFU touch: bump freq; move key from bucket f to f+1.
//
// RUST-SPECIFIC NOTES:
//   - LRU here uses a simple `VecDeque<K>` ordering + `HashMap<K, V>`;
//     O(n) move-to-back. For production use the `lru` crate or hand-roll
//     a doubly-linked list with raw pointers.
//   - LFU uses HashMap<K, (V, freq)> + HashMap<freq, VecDeque<K>>.
//   - std::time::Instant is the monotonic clock.
//
// DRY RUN / EXAMPLE:
//   LRU cap=2: put(1,1); put(2,2); get(1)=Some(1); put(3,3) evicts 2.
//   LFU cap=2: get(1) twice -> freq 3; put(3,3) evicts 2 (freq 1).
//
// COMPLEXITY:
//   This LRU is O(n) per op for simplicity; the LFU is O(1) amortised.

use std::collections::{HashMap, VecDeque};
use std::time::{Duration, Instant};
use std::thread;

pub struct LRUCache {
    cap: usize,
    order: VecDeque<i32>,                  // front = least recent, back = most recent
    map: HashMap<i32, i32>,
}

impl LRUCache {
    pub fn new(cap: usize) -> Self {
        Self { cap, order: VecDeque::new(), map: HashMap::new() }
    }
    fn touch(&mut self, key: i32) {
        if let Some(pos) = self.order.iter().position(|&k| k == key) {
            self.order.remove(pos);
        }
        self.order.push_back(key);
    }
    pub fn get(&mut self, key: i32) -> i32 {
        match self.map.get(&key).copied() {
            Some(v) => { self.touch(key); v }
            None => -1,
        }
    }
    pub fn put(&mut self, key: i32, val: i32) {
        if self.map.contains_key(&key) {
            self.map.insert(key, val);
            self.touch(key);
            return;
        }
        if self.map.len() == self.cap {
            if let Some(evict) = self.order.pop_front() {
                self.map.remove(&evict);
            }
        }
        self.map.insert(key, val);
        self.order.push_back(key);
    }
}

pub struct LFUCache {
    cap: usize,
    min_freq: u32,
    kv: HashMap<i32, (i32, u32)>,                  // key -> (value, freq)
    freq_keys: HashMap<u32, VecDeque<i32>>,        // freq -> LRU within bucket
}

impl LFUCache {
    pub fn new(cap: usize) -> Self {
        Self { cap, min_freq: 0, kv: HashMap::new(), freq_keys: HashMap::new() }
    }
    fn bump(&mut self, key: i32) {
        if let Some((val, freq)) = self.kv.get(&key).copied() {
            // Remove from old bucket
            if let Some(bucket) = self.freq_keys.get_mut(&freq) {
                if let Some(pos) = bucket.iter().position(|&k| k == key) {
                    bucket.remove(pos);
                }
                if bucket.is_empty() {
                    self.freq_keys.remove(&freq);
                    if self.min_freq == freq { self.min_freq += 1; }
                }
            }
            let new_f = freq + 1;
            self.kv.insert(key, (val, new_f));
            self.freq_keys.entry(new_f).or_default().push_back(key);
        }
    }
    pub fn get(&mut self, key: i32) -> i32 {
        if !self.kv.contains_key(&key) { return -1; }
        let val = self.kv[&key].0;
        self.bump(key);
        val
    }
    pub fn put(&mut self, key: i32, val: i32) {
        if self.cap == 0 { return; }
        if self.kv.contains_key(&key) {
            let f = self.kv[&key].1;
            self.kv.insert(key, (val, f));
            self.bump(key);
            return;
        }
        if self.kv.len() >= self.cap {
            if let Some(bucket) = self.freq_keys.get_mut(&self.min_freq) {
                if let Some(evict) = bucket.pop_front() {
                    self.kv.remove(&evict);
                    if bucket.is_empty() { self.freq_keys.remove(&self.min_freq); }
                }
            }
        }
        self.kv.insert(key, (val, 1));
        self.freq_keys.entry(1).or_default().push_back(key);
        self.min_freq = 1;
    }
}

pub struct TTLCache {
    default_ttl: Duration,
    store: HashMap<String, (String, Instant)>,
}

impl TTLCache {
    pub fn new(default_ttl_ms: u64) -> Self {
        Self { default_ttl: Duration::from_millis(default_ttl_ms), store: HashMap::new() }
    }
    pub fn put(&mut self, k: &str, v: &str) {
        self.store.insert(k.to_string(), (v.to_string(), Instant::now() + self.default_ttl));
    }
    pub fn get(&mut self, k: &str) -> Option<String> {
        if let Some((v, dl)) = self.store.get(k) {
            if *dl >= Instant::now() {
                return Some(v.clone());
            }
        }
        self.store.remove(k);
        None
    }
}

fn main() {
    let mut lru = LRUCache::new(2);
    lru.put(1, 1); lru.put(2, 2);
    println!("LRU get(1) = {}", lru.get(1));
    lru.put(3, 3);
    println!("LRU get(2) = {}  (expected -1)", lru.get(2));

    let mut lfu = LFUCache::new(2);
    lfu.put(1, 1); lfu.put(2, 2);
    lfu.get(1); lfu.get(1);
    lfu.put(3, 3);
    println!("LFU get(2) = {}  (expected -1)", lfu.get(2));
    println!("LFU get(1) = {}  (expected 1)", lfu.get(1));
    println!("LFU get(3) = {}  (expected 3)", lfu.get(3));

    let mut ttl = TTLCache::new(50);
    ttl.put("alpha", "value");
    println!("TTL get(alpha) imm.: {:?}", ttl.get("alpha"));
    thread::sleep(Duration::from_millis(60));
    println!("TTL get(alpha) later: {:?}", ttl.get("alpha"));
}

// NOTES
// -----
// Differences from Java:
//   * Rust lacks a built-in ordered hashmap; our LRU uses a VecDeque +
//     HashMap, accepting O(n) touch for simplicity (the `lru` crate offers
//     O(1)).
//   * std::time::Instant is the monotonic clock counterpart of Java's
//     System.nanoTime.
//   * No third-party crates used — pure std for portability.

File: Week 29/cpp/caching.cpp
/*
 * WEEK 29 - C++ ADVANCED TOPICS
 * Topic: Caching - LRU, LFU and TTL
 * File: caching.cpp
 *
 * CONCEPT:
 *   Caches accelerate access to frequently used items by storing them in
 *   limited fast storage. Eviction policy decides what to drop:
 *     - LRU: drop the least recently used (HashMap + doubly-linked list).
 *     - LFU: drop the least frequently used (HashMap + freq buckets).
 *     - TTL: each entry expires after a fixed time.
 *
 * KEY POINTS:
 *   - LRU: O(1) get/put with std::list + std::unordered_map<key, iterator>.
 *   - LFU: O(1) get/put using freq -> list<key>, key -> (value, freq, iter).
 *   - TTL: store deadline; lazily evict on access.
 *
 * ALGORITHM / APPROACH:
 *   LRU: on get/put, move node to list head; if size > cap, drop tail.
 *   LFU: bump freq, move key from bucket f to bucket f+1; evict from
 *        minFreq bucket's tail when full.
 *   TTL: compare std::chrono::steady_clock::now() against entry's expiry.
 *
 * C++-SPECIFIC NOTES:
 *   - std::list::iterator stays valid on insert/erase elsewhere.
 *   - std::chrono::steady_clock for monotonic time.
 *
 * DRY RUN / EXAMPLE:
 *   LRU cap=2: put(1,1); put(2,2); get(1)=1; put(3,3) evicts 2; get(2)=-1.
 *   LFU cap=2: put(1); put(2); get(1) twice -> evict 2 on put(3).
 *
 * COMPLEXITY:
 *   LRU/LFU: O(1) per op.
 *   TTL:     O(1) per op.
 */

#include <iostream>
#include <list>
#include <unordered_map>
#include <chrono>
#include <thread>
#include <string>

using namespace std;

// ---------------------------------------------------------------------------
// LRU Cache
// ---------------------------------------------------------------------------
class LRUCache {
    int cap;
    list<pair<int,int>> lst; // (key, value); front = most recent
    unordered_map<int, list<pair<int,int>>::iterator> mp;
public:
    explicit LRUCache(int c) : cap(c) {}

    int get(int key) {
        auto it = mp.find(key);
        if (it == mp.end()) return -1;
        lst.splice(lst.begin(), lst, it->second);
        return it->second->second;
    }

    void put(int key, int val) {
        auto it = mp.find(key);
        if (it != mp.end()) {
            it->second->second = val;
            lst.splice(lst.begin(), lst, it->second);
            return;
        }
        if ((int)lst.size() == cap) {
            mp.erase(lst.back().first);
            lst.pop_back();
        }
        lst.emplace_front(key, val);
        mp[key] = lst.begin();
    }
};

// ---------------------------------------------------------------------------
// LFU Cache (O(1) with freq buckets)
// ---------------------------------------------------------------------------
class LFUCache {
    int cap, minFreq;
    unordered_map<int, pair<int,int>> kv;             // key -> (value, freq)
    unordered_map<int, list<int>> fl;                 // freq -> list of keys (LRU within)
    unordered_map<int, list<int>::iterator> kit;      // key -> iterator in fl[freq]

    void bump(int key) {
        auto& [val, freq] = kv[key];
        fl[freq].erase(kit[key]);
        if (fl[freq].empty()) {
            fl.erase(freq);
            if (minFreq == freq) ++minFreq;
        }
        ++freq;
        fl[freq].push_front(key);
        kit[key] = fl[freq].begin();
    }
public:
    explicit LFUCache(int c) : cap(c), minFreq(0) {}

    int get(int key) {
        if (!kv.count(key)) return -1;
        bump(key);
        return kv[key].first;
    }

    void put(int key, int val) {
        if (cap <= 0) return;
        if (kv.count(key)) {
            kv[key].first = val;
            bump(key);
            return;
        }
        if ((int)kv.size() >= cap) {
            int evict = fl[minFreq].back();
            fl[minFreq].pop_back();
            if (fl[minFreq].empty()) fl.erase(minFreq);
            kv.erase(evict);
            kit.erase(evict);
        }
        kv[key] = {val, 1};
        fl[1].push_front(key);
        kit[key] = fl[1].begin();
        minFreq = 1;
    }
};

// ---------------------------------------------------------------------------
// TTL Cache
// ---------------------------------------------------------------------------
class TTLCache {
    using Clock = chrono::steady_clock;
    chrono::milliseconds defTtl;
    unordered_map<string, pair<string, Clock::time_point>> store;
public:
    explicit TTLCache(int defaultTtlMs) : defTtl(defaultTtlMs) {}
    void put(const string& k, const string& v) {
        store[k] = {v, Clock::now() + defTtl};
    }
    bool get(const string& k, string& out) {
        auto it = store.find(k);
        if (it == store.end()) return false;
        if (it->second.second < Clock::now()) { store.erase(it); return false; }
        out = it->second.first;
        return true;
    }
};

int main() {
    LRUCache lru(2);
    lru.put(1, 1); lru.put(2, 2);
    cout << "LRU get(1) = " << lru.get(1) << "\n";
    lru.put(3, 3);
    cout << "LRU get(2) = " << lru.get(2) << "  (expected -1)\n";

    LFUCache lfu(2);
    lfu.put(1, 1); lfu.put(2, 2);
    lfu.get(1); lfu.get(1);
    lfu.put(3, 3);
    cout << "LFU get(2) = " << lfu.get(2) << "  (expected -1)\n";
    cout << "LFU get(1) = " << lfu.get(1) << "  (expected 1)\n";
    cout << "LFU get(3) = " << lfu.get(3) << "  (expected 3)\n";

    TTLCache ttl(50);
    ttl.put("alpha", "value");
    string out;
    cout << "TTL get(alpha) imm.: " << (ttl.get("alpha", out) ? out : "MISS") << "\n";
    this_thread::sleep_for(chrono::milliseconds(60));
    cout << "TTL get(alpha) later: " << (ttl.get("alpha", out) ? out : "MISS") << "\n";
    return 0;
}

/*
 * NOTES
 * -----
 * Differences from Java:
 *   - C++ uses std::list + iterator stability instead of a hand-rolled
 *     doubly-linked list.
 *   - LFU here is O(1); Java's system_design.java demoes only LRU.
 *   - std::chrono::steady_clock is the C++ monotonic clock.
 */
